demos/guide/tls-server-block.c

```diff
@@ -0,0 +1,233 @@
+ *  https://www.openssl.org/source/license.html
+ */
+
+/*
+ * NB: Changes to this file should also be reflected in
+ * doc/man7/ossl-guide-tls-client-non-block.pod
+ */
+
+#include <err.h>
+#include <string.h>
+
+/* Include the appropriate header file for SOCK_STREAM */
+#ifdef _WIN32 /* Windows */
+# include <winsock2.h>
+#else /* Linux/Unix */
+# include <sys/socket.h>
+# include <sys/select.h>
+#endif
+
+#include <openssl/bio.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+static const char cache_id[] = "OpenSSL Demo Server";
+
+/* Minimal TLS echo server. */
+int main(int argc, char *argv[])
+{
+    int res = EXIT_FAILURE;
+    long opts;
+    const char *hostport;
+    SSL_CTX *ctx = NULL;
+    BIO *acceptor_bio;
+
+    if (argc != 2) {
+        errx(res, "Usage: %s [host:]port", argv[0]);
+    }
+    hostport = argv[1];
+
+    /*
+     * An SSL_CTX holds shared configuration information for multiple
+     * subsequent per-client SSL connections.
+     */
+    ctx = SSL_CTX_new(TLS_server_method());
+    if (ctx == NULL) {
+        ERR_print_errors_fp(stderr);
+        errx(res, "Failed to create server SSL_CTX");
+    }
+
+    /*
+     * TLs versions older than TLS 1.2 are deprecated by IETF and SHOULD
+     * be avoided if possible.
+     */
+    if (!SSL_CTX_set_min_proto_version(ctx, TLS1_2_VERSION)) {
+        SSL_CTX_free(ctx);
+        ERR_print_errors_fp(stderr);
+        errx(res, "Failed to set the minimum TLS protocol version");
+    }
+
+#if 0
+    /*
+     * In applications (e.g. SMTP) where most clients are performing
+     * unauthenticated opportunistic TLS it may make sense to set the security
+     * level to 0, allowing weaker encryption parameters, which are still
+     * stronger that a potential cleartext fallback.
+     *
+     * The default security level is 1, which is roughly equivalent to that of
+     * 80 bit symmetric keys, or 1024-bit RSA or finite-field Diffie-Hellman
+     * keys.  Notably, security level 1 no longer allows the use of SHA-1 in
+     * certificate signatures, key exchange or in the TLS 1.[01] PRF.
+     */
+    SSL_CTX_set_security_level(ctx, 0);
+#endif
+
+    /*
+     * Tolerate clients hanging up without a TLS "shutdown".  Appropriate in all
+     * application protocols which perform their own message "framing", and
+     * don't rely on TLS to defend against "truncation" attacks.
+     */
+    opts = SSL_OP_IGNORE_UNEXPECTED_EOF;
+
+    /*
+     * Block potential CPU-exhaustion attacks by clients that request frequent
+     * renegotiation.  This is of course only effective if there are existing
+     * limits on initial full TLS handshake or connection rates.
+     */
+    opts |= SSL_OP_NO_RENEGOTIATION;
+
+    /* Apply the selection options */
+    SSL_CTX_set_options(ctx, opts);
+
+    /*
+     * Most servers elect to use their own cipher preference rather than that of
+     * the client.
+     */
+    SSL_CTX_set_options(ctx, SSL_OP_CIPHER_SERVER_PREFERENCE);
```

Comment:
Why two separate calls to `SSL_CTX_set_options` rather than adding it to `opts` like you did for the two other options above? Seems inconsistent